libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedModulePath.java

```diff
@@ -0,0 +1,363 @@
+            }
+            return Optional.empty();
+        }
+
+        String pn = name.substring(0, index).replace('/', '.');
+        if (isPackageName(pn)) {
+            return Optional.of(pn);
+        } else {
+            // not a valid package name
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Returns an optional containing the service name from a given services path name, or an
+     * empty optional if none.
+     */
+    static Optional<String> toServiceName(String cf) {
+        if (cf.startsWith(SERVICES_PREFIX) == false) {
+            throw new IllegalArgumentException();
+        }
+        int index = cf.lastIndexOf("/") + 1;
+        if (index < cf.length()) {
+            String prefix = cf.substring(0, index);
+            if (prefix.equals(SERVICES_PREFIX)) {
+                String sn = cf.substring(index);
+                if (isClassName(sn)) {
+                    return Optional.of(sn);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    /** Returns the next line from the reader, skipping comment lines. */
+    static String nextLine(BufferedReader reader) throws IOException {
+        String ln = reader.readLine();
+        if (ln != null) {
+            int ci = ln.indexOf('#');
+            if (ci >= 0) {
+                ln = ln.substring(0, ci);
+            }
+            ln = ln.trim();
+        }
+        return ln;
+    }
+
+    static boolean isPackageName(String name) {
+        return isTypeName(name);
+    }
+
+    static boolean isClassName(String name) {
+        return isTypeName(name);
+    }
+
+    static boolean isTypeName(String name) {
+        int next;
+        int off = 0;
+        while ((next = name.indexOf('.', off)) != -1) {
+            String id = name.substring(off, next);
+            if (isJavaIdentifier(id) == false) {
+                return false;
+            }
+            off = next + 1;
+        }
+        String last = name.substring(off);
+        return isJavaIdentifier(last);
+    }
+
+    static boolean isJavaIdentifier(String str) {
+        if (str.isEmpty()) {
+            return false;
+        }
+
+        int first = Character.codePointAt(str, 0);
+        if (Character.isJavaIdentifierStart(first) == false) {
+            return false;
+        }
+
+        int i = Character.charCount(first);
+        while (i < str.length()) {
+            int cp = Character.codePointAt(str, i);
+            if (Character.isJavaIdentifierPart(cp) == false) {
+                return false;
+            }
+            i += Character.charCount(cp);
+        }
+        return true;
+    }
+
+    private static final String MANIFEST_PATH = "META-INF/MANIFEST.MF";
+
+    static String moduleNameFromManifestOrNull(Path path) {
+        Path mp = path.resolve(MANIFEST_PATH);
+        if (Files.exists((mp))) {
+            try (InputStream is = Files.newInputStream(mp)) {
+                if (is != null) {
```

Comment:
When can newInputStream return null? I don't see it in the javadocs as a valid return